lgc/patch/PatchBufferOp.cpp

```diff
@@ -1475,6 +1475,71 @@ static unsigned getLSIntrinsicID(bool isLoad, bool isInvariant, unsigned accessS
   return Intrinsic::amdgcn_raw_buffer_store;
 }
 
+// =====================================================================================================================
+// Recalculate offset by moving the constant part of the offset from "Value" component to "unsigned" component.
+//
+// Analyze the GEP contributing to "Value" component, and move the constant part to the "unsigned" component.
+//
+// @param initialOffset : The initial offset, consisting of (Value, unsigned) components.
+// @returns : The updated offset, where some of "Value" component has been moved to "unsigned" component.
+std::pair<Value *, unsigned> PatchBufferOp::recalculateOffset(std::pair<Value *, unsigned> initialOffset) {
+
+  auto ptrToInt = dyn_cast<PtrToIntInst>(initialOffset.first);
+  if (!ptrToInt)
+    return initialOffset;
+
+  auto bitcast = dyn_cast<BitCastInst>(ptrToInt->getPointerOperand());
+  Value *gepCand = bitcast ? bitcast->getOperand(0) : ptrToInt->getPointerOperand();
+
+  auto GEP = dyn_cast<GetElementPtrInst>(gepCand);
+  if (!GEP)
+    return initialOffset;
+
+  // Find the splitting point in GEP: first const index in the series of all const indices.
+  auto idxEnd = GEP->idx_end();
+  auto idxFirstConstInSequence = idxEnd;
+
+  for (auto idxIt = GEP->idx_begin(); idxIt != idxEnd; ++idxIt) {
+    if (!isa<ConstantInt>(*idxIt)) {
+      idxFirstConstInSequence = idxIt + 1;
+    }
+  }
+
+  if (idxFirstConstInSequence == idxEnd)
+    return initialOffset;
+
+  // Prepare a list of GEP indices containing the const indices to query the total offset accumulated.
+  SmallVector<Value *, 8> frontIndices(GEP->idx_begin(), idxFirstConstInSequence);
+  auto constSequenceGEPType = GetElementPtrInst::getIndexedType(GEP->getSourceElementType(), frontIndices);
+
+  SmallVector<Value *, 8> constSequenceIndices;
+  constSequenceIndices.push_back(m_builder->getInt32(0));
+  constSequenceIndices.insert(constSequenceIndices.end(), idxFirstConstInSequence, idxEnd);
+
+  const DataLayout &DL = m_builder->GetInsertBlock()->getModule()->getDataLayout();
+  APInt Offset(DL.getIndexSizeInBits(GEP->getPointerAddressSpace()), 0);
+  if (!GEPOperator::accumulateConstantOffset(constSequenceGEPType, constSequenceIndices, DL, Offset))
+    return initialOffset;
+
+  int64_t newOffsetImm = initialOffset.second + Offset.getSExtValue();
+  if (newOffsetImm < 0 || newOffsetImm > UINT_MAX)
+    return initialOffset;
+
+  // Create a new GEP, without the sequence of constants at the end. The accumulated constant part of the GEP
+  // will be moved to the the "unsigned" component.
+  m_builder->SetInsertPoint(GEP);
+
+  Value *const pointer = getPointerOperandAsInst(GEP->getPointerOperand());
+  SmallVector<Value *, 8> newIndices(GEP->idx_begin(), idxFirstConstInSequence);
```

Comment:
Redundant with frontIndices. Also, could/should probably be an ArrayRef. Also, what about the case where frontIndices could be empty because all indices are constant?

Reply:
All constant indices would've been optimized by now, but I will make sure that that case is handled, just to err on the side of caution.

Reply:
Handled the empty frontIndices earlier in the code by exiting early and commoned up frontIndices. I did not figure out how to use ArrayRef here as I also want to pass data to createGEP later in the code.